### ISSUE TITLE
自定义多轮对话规划器实现问题

### ISSUE THREAD

**Issue by LuckyDoggy** [modelscope/ms-swift#6187]
自定义多轮对话规划器和奖励函数，代码如下
```python
import asyncio
import os
import re
import textwrap
from collections import Counter
from copy import deepcopy
from typing import Dict, List, Union

import json
import torch

from swift.llm import PtEngine, RequestConfig, RolloutInferRequest, Template, to_device
from swift.llm.infer.protocol import ChatCompletionResponse, ChatCompletionResponseChoice
from swift.plugin import ORM, orms, rm_plugins
# register context manager(used in gym training)
from swift.plugin.context_manager import ContextManager, context_managers
from swift.plugin.env import Env, envs
from swift.plugin.multi_turn import MultiTurnScheduler, multi_turns
from swift.plugin.rm_plugin import DefaultRMPlugin
from swift.utils import get_logger

#===================================vision-sr1==================================
caption_reasoner_prompt_template = "Text description: [vision_caption]\nQuestion: [QUESTION]\n\nYou are provided a text description of a problem and a question. Determine the answer to the question based on the text description. First provide an internal step-by-step reasoning within <think> </think> tags, then provide a single word or phrase answer in \\boxed{}"



def extract_description(predict: str):
    match = re.search(r"<description>([\s\S]*?)</description>", predict, re.DOTALL)
    if not match:
        return None
    return match.group(1).strip()

def parser_question(original_user_prompt):
    question = original_user_prompt.split("Question: ")[1].split("\n\n")[0].strip()
    return question


def generate_new_prompt(completion: str, original_user_prompt: str):
    """
    解析第一轮回复，构造第二轮prompt
    """

    description = extract_description(completion)
    if description is None:
        return None
    question = parser_question(original_user_prompt)
    user_query = caption_reasoner_prompt_template.replace("[QUESTION]", question)
    user_query = user_query.replace("[vision_caption]", description)    
    # 默认返回一个接力 Prompt
    return user_query


class SelfRewardScheduler(MultiTurnScheduler):


    async def run(self, infer_request: 'RolloutInferRequest', request_config: 'RequestConfig',
                  **kwargs) -> List['RolloutOutput']:
        from swift.llm.infer.protocol import RolloutOutput
        current_request_turn1 = deepcopy(infer_request)
        
        # --------------------- 第一轮：多模态推理（核心训练样本） ---------------------
        current_turn = 1
        messages_turn1 = current_request_turn1.messages
        # print(f"messages_turn1: {messages_turn1}")
        original_user_prompt = messages_turn1[1]['content']
        
        # 1.1 模型推理
        response_turn1: 'ChatCompletionResponse' = await self.infer_engine.infer_async(
            current_request_turn1, request_config, **kwargs)
        response_choice_turn1: 'ChatCompletionResponseChoice' = response_turn1.choices[0]
        completion_turn1 = response_choice_turn1.message.content
        
        # 1.2 记录第一轮历史（用于 RolloutOutput.messages）
        messages_full_history_turn1 = deepcopy(messages_turn1)
        messages_full_history_turn1.append({'role': 'assistant', 'content': completion_turn1})
        
        response_token_ids_t1 = response_choice_turn1.token_ids
        
 
        # --------------------- 第二轮：纯文本推理（只为奖励计算而推理） ---------------------
        
        # 2.1 尝试解析，获取下一轮 Prompt
        new_prompt = generate_new_prompt(completion_turn1, original_user_prompt)
        # 2.1 校验第一轮输出格式，构造第二轮prompt
        if new_prompt is not None:
            # 2.2 重置请求，创建第二轮推理请求
            current_request_turn2 = deepcopy(infer_request) 
            messages_turn2_input = [{'role': 'user', 'content': new_prompt}]
            current_request_turn2.messages = messages_turn2_input
            
            # 2.3 移除多模态数据！
            if hasattr(current_request_turn2, 'images'): current_request_turn2.images = []
            if hasattr(current_request_turn2, 'videos'): current_request_turn2.videos = []
            if current_request_turn2.data_dict:
                 if 'images' in current_request_turn2.data_dict: current_request_turn2.data_dict['images'] = []
                 if 'videos' in current_request_turn2.data_dict: current_request_turn2.data_dict['videos'] = []

            # 2.4 模型推理
            response_turn2: 'ChatCompletionResponse' = await self.infer_engine.infer_async(
                current_request_turn2, request_config, **kwargs)
            response_choice_turn2: 'ChatCompletionResponseChoice' = response_turn2.choices[0]
            completion_turn2 = response_choice_turn2.message.content
        
        # --------------------- 最终RolloutOutput 构造 ---------------------
     
        # 构造最终的 RolloutOutput（只返回第一轮以及第二轮的response）
        # print(f"messages: {messages_full_history_turn1}")
        # print(f"response_token_ids: {response_token_ids_t1}")
        output_turn1 = RolloutOutput(
            response=response_turn1,
            messages=messages_full_history_turn1, 
            response_token_ids=response_token_ids_t1,
            rollout_infos={
                'num_turns': current_turn,
                'turn2_executed': new_prompt is not None, # 记录第二轮是否执行
                'see_think_response': completion_turn1,
                'caption_reasonser_response': response_choice_turn2.message.content if new_prompt is not None else ""
            }
        )
        
        # 只返回第一轮的 RolloutOutput 列表
        return [output_turn1]

    # 覆盖 step 和 check_finished，防止被意外调用
    def step(self, infer_request: 'RolloutInferRequest', response_choice: 'ChatCompletionResponseChoice',
             current_turn: int) -> Dict:
        raise NotImplementedError("TwoTurnVideoScheduler uses a custom run() and does not rely on step().")

    def check_finished(self, infer_request: 'RolloutInferRequest', response_choice: 'ChatCompletionResponseChoice',
                       current_turn: int) -> bool:
        return True # 终止逻辑在 run() 中处理



multi_turns['self_reward_scheduler'] = SelfRewardScheduler

def format_reward(predict: str) -> float:
    pattern = re.compile(
        r"^\s*<description>.*?</description>\s*"    # the image description block
        r"<think>.*?</think>\s*"                   # the reasoning block
        r"\\boxed\{.*?\}\s*$",                     # the final answer
        re.DOTALL
    )
    return 1.0 if pattern.fullmatch(predict) else 0.0

class SeeThinkFormatReward(ORM):
    def __call__(self, completions, **kwargs):
        infos = kwargs.get('rollout_infos', {})
        rewards = [format_reward(c) for c in completions]
        return rewards
        

# class SeeThinkFormatReward():
#     def __call__(self, completions, **kwargs):
#         infos = kwargs.get('rollout_infos', {})

orms['see_think_format_reward']= SeeThinkFormatReward
```

rollout脚本
```bash
export DECORD_EOF_RETRY_MAX=20480

model_kwargs='{"fps": 2.0, "fps_max_frames": 360, "video_total_pixels": 18063360, "video_min_pixels": 50176}'


DECORD_EOF_RETRY_MAX=20480 \
CUDA_VISIBLE_DEVICES=4,5,6,7 \
swift rollout \
  --model /workspace/Qwen2.5-VL-7B-Instruct/ \
  --vllm_use_async_engine true \
  --external_plugins /workspace/mount/b100_zaip_data/yjj/projects/swift_vision_sr1/plugin/plugin.py \
  --multi_turn_scheduler self_reward_scheduler \
  --vllm_max_model_len 27136 \
  --vllm_tensor_parallel_size 1 \
  --vllm_data_parallel_size 4 \
  --port 8000 \
  --model_kwargs "$model_kwargs"
```
训练脚本
```bash
MASTER_PORT=29501 \
DECORD_EOF_RETRY_MAX=20480 \
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model /workspace/Qwen2.5-VL-7B-Instruct/ \
    --add_version false \
    --reward_funcs see_think_format_reward \
    --reward_weights 1.0 \
    --train_type full \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_host 127.0.0.1 \
    --vllm_server_port 8000 \
    --vllm_server_pass_dataset true \
    --torch_dtype bfloat16 \
    --dataset XXXX \
    --max_completion_length 4096 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 1 \
    --split_dataset_ratio 0 \
    --save_steps 100 \
    --save_total_limit 500 \
    --logging_steps 1 \
    --max_length 27136 \
    --output_dir xxxx \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --dataset_num_proc 4 \
    --num_generations 4 \
    --temperature 1.0 \
    --num_iterations 1 \
    --deepspeed zero3 \
    --attn_impl flash_attn \
    --log_completions true \
    --use_liger_kernel True \
    --external_plugins /xxxx/plugin.py \
    --model_kwargs "$model_kwargs"
```

报错信息如下：

<img width="1395" height="835" alt="Image" src="https://github.com/user-attachments/assets/bef26407-90fc-4800-b50d-1e8d014f8a3e" />

**Comment by hjh0119** [comment]
please retry with --use_liger_kernel false

**Comment by LuckyDoggy** [comment]
> please retry with --use_liger_kernel false

Thank you, it works. Why can't I use liger kernel?

**Comment by hjh0119** [comment]
The Liger loss is implemented in single-turn GRPO via Liger-Kernel, but is not currently compatible with multi-turn scenarios.
